### PR TITLE
Feature: Adds APIs necessary for supporting consignor offer responses

### PR DESCRIPTION
**Pull request by sweir27**
The overarching goal of this is to pave the way for consignors to be able to respond to offers from a form on artsy.net. 

See previous modeling PR [here](https://github.com/artsy/convection/pull/844).

This PR:
1. Updates the `offer_resolver` to allow users to retrieve their own offers.
2. Adds a new mutation, `createConsignmentOfferResponse` to allow users to create responses to offers.

Screen Shot 2020-11-11 at 8 54 01 PM

I've left some more comments inline!

### CODE REVIEW

File: app/graphql/mutations/create_offer_response_mutation.rb
Comment (sweir27) [inline app/graphql/mutations/create_offer_response_mutation.rb:10]: The input type here could be updated to match the database requirements. I worried that that would force us to keep multiple types up-to-date (and especially because the rejection reasons themselves are longer strings), so opted to rely on the ActiveRecord validation instead.

File: app/graphql/mutations/create_offer_response_mutation.rb
Comment (sweir27) [inline app/graphql/mutations/create_offer_response_mutation.rb:12]: Externally, these types are prefaced with `consignment_` to disambiguate from BNMO offers (I'm guessing! I was also following convention here).

File: app/graphql/resolvers/create_offer_response_resolver.rb
Comment (sweir27) [inline app/graphql/resolvers/create_offer_response_resolver.rb:5]: To keep it simple, this requires that a user ID is present in the JWT. Further authorization is done below.

File: app/graphql/resolvers/create_offer_response_resolver.rb
Comment (sweir27) [inline app/graphql/resolvers/create_offer_response_resolver.rb:16]: Ensure that the offer they're responding to is indeed their own.

File: app/graphql/resolvers/offer_resolver.rb
Comment (sweir27) [inline app/graphql/resolvers/offer_resolver.rb:12]: This basically says-- if the requesting JWT includes a `partner` (or `admin`) role, and you've passed in a `gravity_partner_id` argument, _then_ return just the partner's offer that matches the request. 

Otherwise we assume you're a user or admin looking for a single offer.

This means that the `gravity_partner_id` input is responsible for determining this scope.

File: app/graphql/types/query_type.rb
Comment (sweir27) [inline app/graphql/types/query_type.rb:10]: Because we can now return offers for a user, this is no longer required.

File: spec/requests/api/graphql/queries/offer_spec.rb
Comment (sweir27) [inline spec/requests/api/graphql/queries/offer_spec.rb:68]: Changed the message here in order to not leak that an offer exists if you're not authorized to view it.

File: spec/requests/api/graphql/queries/offer_spec.rb
Comment (sweir27) [inline spec/requests/api/graphql/queries/offer_spec.rb:236]: We don't allow users to access offers in the draft state. I _think_ this is okay, but should we restrict other states too?

File: spec/requests/api/graphql/mutations/create_consignment_offer_response_spec.rb
Comment (pepopowitz) [inline spec/requests/api/graphql/mutations/create_consignment_offer_response_spec.rb:39]: [non-blocking] Another useful test here would be `context 'with someone else's offer'` .. `it 'returns an error for that request'`. It _looks_ like that's accounted for in the underlying mutation, but it'd be nice to have a test to prove it. 

I would merge without this test but if you have the time & feel like it's easy to add before merging, it's a welcome addition.

File: spec/requests/api/graphql/queries/offer_spec.rb
Comment (pepopowitz) [reply spec/requests/api/graphql/queries/offer_spec.rb:236]: These are the states on an offer: 

```
  STATES = [
    DRAFT = 'draft',
    SENT = 'sent',
    ACCEPTED = 'accepted',
    REJECTED = 'rejected',
    LAPSED = 'lapsed',
    REVIEW = 'review'
  ].freeze
```

Aside from "draft", all of them seem like states that, if I were a consignor, I might want to reach out about if I'd changed my mind or had questions. I feel like we're good allowing responses in any states except "draft".

**Review by pepopowitz — APPROVED** [review]
❤️ ❤️ ❤️ 

This is amazing! 

I left a comment about one test that seems missing on creating an offer response. If you've got the time & it's easy to add, it'd be great! Let me know if you don't, and I'll merge this and add a test after.

File: app/graphql/mutations/create_offer_response_mutation.rb
Comment (jonallured) [reply app/graphql/mutations/create_offer_response_mutation.rb:10]: **question [non-blocking]:** minimum viable type checking

Can you elaborate on your worry about marking this as required? Seems like it would be better error feedback for one to pass an invalid string here than to leave it off and then get an error about the thing being required, no? I think what you're getting at is that you _could_ have created an enum with the possible valid values, but that seemed extra so you didn't. I guess my feeling is that marking this as a required string is a good middle ground. 🤷‍♂️ 

*format via [Conventional Comments](https://conventionalcomments.org/)*

File: app/graphql/mutations/create_offer_response_mutation.rb
Comment (jonallured) [reply app/graphql/mutations/create_offer_response_mutation.rb:12]: Naming is hard! 😝

File: app/graphql/resolvers/create_offer_response_resolver.rb
Comment (jonallured) [reply app/graphql/resolvers/create_offer_response_resolver.rb:16]: **question [non-blocking]:** check validity of intent sooner?

What happens if one tries to respond to an offer where they are not the gravity user? I wonder if this check belongs in the `valid?` method? 🤔 

*format via [Conventional Comments](https://conventionalcomments.org/)*

**Comment by jonallured** [comment]
Some random thoughts but this is super cool and I'm embarrassed to say that I did not know we were still using Google forms for the consignor's offer responses!! 😝

File: app/graphql/resolvers/create_offer_response_resolver.rb
Comment (sweir27) [reply app/graphql/resolvers/create_offer_response_resolver.rb:16]: That check is just below!

Actually perhaps a question for you-- in the resolvers, I see a mix of approaches (some validations in `valid?` and others within the resolvers themselves).

I tried to keep the `valid?` methods fairly lean and based off of attributes that we get "for free" (i.e. whether you're an admin, user, partner, etc.) and to leave the more complex checks (for ownership, etc.) to the resolver method.

Many ways to approach this, curious if you have a strong opinion?

File: app/graphql/mutations/create_offer_response_mutation.rb
Comment (sweir27) [reply app/graphql/mutations/create_offer_response_mutation.rb:10]: Ah! The field is _not_ required, I just opted to not create an additional Enum type to capture the potential inputs given they'll also be checked at creation time.

But this is not a strong opinion, I can easily create an Enum to give quicker feedback (which in retrospect might be a huge help to the UI builder).

Comment (jonallured) [reply app/graphql/mutations/create_offer_response_mutation.rb:10]: OMG Jon read the code my dude. What you have here totally makes sense. I saw that validation check and assumed it was required. I see now that it's NOT required and we only validate that if present, it be a certain set of values.

In terms of building the UI, totally up to you if this enum helps. Could totally come in a follow up if you wanna keep moving.

File: app/graphql/resolvers/create_offer_response_resolver.rb
Comment (jonallured) [reply app/graphql/resolvers/create_offer_response_resolver.rb:16]: Yeah I just read through the resolvers and you're right that there are various approaches. I think we are aligned on the line between `valid?` doing some sanity checking and the `run` method raising errors. Where I land on this is that `valid?` should probably stay out of the database. So we can check things like whether the token includes the admin flag or if the combination of query args are allowed, whatever. But checking Rails-level validations or doing something that hits the db is probably out of scope for the valid method.

So given that, I retract my question, haha. What you have here looks wonderful! 🙌

**Pull request merged**
